1334/activity orders from api rebased

alfetopito opened the pull request:
# Summary

Closes #1334 

Follows up/supersedes https://github.com/gnosis/cowswap/pull/1395

Same content as that, but rebased against `develop` after the `presign` merge.

Lot's of changes, some things might not work
Activity orders from API: "Your orders, wherever you are"

Main changes:
- At start/network/account change, fetches orders from API
- Shows only orders for connected account/network
- Shows only txs for connected account/network
- Displays at most 10 regulars orders, and as many pending as there are
- Shows orders sorted by creation date, descending
- Replaced `Clear activity` with `View all orders`
- Added a `View all orders` at the bottom of activity list
- Show orders that have Tokens not yet loaded in the current UI

# To Test

1. Connect a wallet with existing orders
2. Open Account Activities sidebar
* There should be 10 orders or less, doesn't matter how old they are (There might be more if you have more than 10 pending orders)
* There should be a link on top right that takes you to the Explorer to see all of your orders
3. Change account AND/OR network
* Orders should be updated to match the selected account/network
4. Place an order
5. While it's pending, open the same page in another browser / device (:warning: important to test in a different browser otherwise the same local storage will be used)
* The pending order should be visible as pending in the new browser / device


# *NOT* part of this PR

- Load txs the same way as orders
- Refresh pending orders in the background - right now they are updated only on account/network change

# Not yet addressed

- [x] `presignature pending` state on order insertion before calling reducer
- [x] background update of orders for not the current connected wallet (https://github.com/gnosis/cowswap/pull/1684)
- [x] issues with safe not reflecting pre-signing
- [x] random safe hard errors

Commit:
Added /account/<address>/orders api endpoint

Commit:
Added helper function to get explorer address link

Commit:
Requiring needed properties instead of concrete types on utils fns

Commit:
Added addOrUpdateOrdersBatch to orders/actions

Commit:
Added useAddOrUpdateOrdersBatch to orders/hooks

Commit:
Added addOrUpdateOrdersBatch to orders/reducer

Commit:
Added APIOrdersUpdater that checks for orders from API on connection

Commit:
Adjusted displayed orders:

- At most 10 regular orders and as much pending as there are
- Show only connected account orders
- Sort orders by creation date descending

Commit:
Show only txs for connected account

Commit:
Refactoring

Commit:
Don't actually need a helper fn...

Commit:
Added comment to make sort more explicit

Commit:
Refactoring

Commit:
Refactor: cache tokens loaded const

Commit:
Refactor: removed duplicated css property

Commit:
Refactor: removed duplicated CSS property

Commit:
Refactor: extracted function transformApiOrderToStoreOrder

Commit:
Refactor: renamed _orders to apiOrders

Commit:
Refactor: removed duplicated CSS property

Commit:
Adjusting `View all orders` link size

Commit:
Added `View all orders` to the bottom of account activity

Commit:
Refactored computeOrderSummary to deal with pending orders

Commit:
Computing order summary for orders loaded from API

Commit:
Setting isCancelling state if order is pending and cancelled on API

Commit:
Actually syncing `isCancelling` flag for existing orders

Commit:
Refactor: created const for magic number MAXIMUM_ORDERS_TO_DISPLAY

Commit:
Refactor: simplified addOrUpdateOrdersBatch reducer

Commit:
Refactor: renamed API -> Api on file name

Commit:
Refactor: replaced magic number with AMOUNT_OF_ORDERS_TO_FETCH

Commit:
Refactor: converted classifyLocalStatus fn to a map

Commit:
Refactor: removed `Batch` from addOrUpdateOrders

Commit:
Refactor: moved error exit condition up

alfetopito committed:
1334/load new tokens not yet loaded (#1504)

* Added non-hook methods for dealing with contracts

* New hook useTokensLazy

* Exporting stuff from original codebase

* Added getMultipleCallsResults under custom/state/multicall/utils

Non-hook way of fetching multiple onchain queries through multicall

* Refactor ApiOrdersUpdater to load tokens not yet in the UI

* Typo fix

* 1334/do not refetch when token added (#1536)

* Removing annoying log

* No longer re-fetching orders when a token is added to the list

Co-authored-by: Leandro Boscariol <leandro.boscariol@gnosis.io>

* Refactored useTokensLazy:

- Added contracts cache
- Improved comments

* Improved doc string on multicall utils

Co-authored-by: Leandro Boscariol <leandro.boscariol@gnosis.io>

Commit:
Using 'Link to Explorer' instead of 'Clear activities'

Commit:
Fix new Erc20 import path

Commit:
Adjust to new multicall interface and fetchChunk changes

Commit:
Adapt to renamed of ApiOrderStatus to OrderTransitionStatus

Commit:
Adjusting (again) styles for `view all activity`

Commit:
Handling presignaturePending state from backend

Commit:
The order is concrete at this point, no need for `?.`

MareenG commented:
Hey @alfetopito  , 
I tried to delete `redux_localstorage_simple_orders` several times. However, it doesn't change anything and remains in the local storage.

elena-zh commented:
Hey @alfetopito , I have just noticed that 'Valid to' date is not replaced with 'Filled on' for filled orders. 
![image](https://user-images.githubusercontent.com/70885163/139697348-dcb8d84d-91e5-47c0-a83c-9beb04df6a16.png)

However, I'm not sure if it is connected with the current PR. 
Please let me know if I need to create a separate issue for this.

alfetopito commented:
> Hey @alfetopito , I tried to delete `redux_localstorage_simple_orders` several times. However, it doesn't change anything and remains in the local storage.

@MareenG 
How have you tried to delete?
Usually, selecting the key and clicking the `delete` key on the keyboard should do it.
Otherwise you can right click the key and select `Delete`.
Also, the app recreates it right away, so maybe delete and refresh?

alfetopito commented:
> Hey @alfetopito , I have just noticed that 'Valid to' date is not replaced with 'Filled on' for filled orders. ![image](https://user-images.githubusercontent.com/70885163/139697348-dcb8d84d-91e5-47c0-a83c-9beb04df6a16.png)
> 
> However, I'm not sure if it is connected with the current PR. Please let me know if I need to create a separate issue for this.

@elena-zh that unfortunately is a "won't fix" for now.
The API doesn't have a fulfillment date, so "Filled on" will only be visible in the same browser where the order was placed.

MareenG commented:
> @MareenG How have you tried to delete? Usually, selecting the key and clicking the `delete` key on the keyboard should do it. Otherwise you can right click the key and select `Delete`. Also, the app recreates it right away, so maybe delete and refresh?

That Is exactly how I did it yesterday. I might have phrased it wrongly, sorry. It deleted the `redux_localstorage_simple_orders` but the orders remained in the orders activities (Still showing xDai trades when mainnet is selected).  
However I just tried it again and today it worked. Before doing it the orders where still displayed in the wrong network. After deleting it, it was now as expected 🤷 . I placed a few orders in different networks and everything was ok.
I will do a few more trades over the day and also do it in FF. I will let you know in case this shows up again :)

alfetopito committed:
1334/only update connected wallet (#1684)

* Only update cancelled orders for connected wallet

* Only check for unfillable orders for connected wallet

* Only check pending orders for connected wallet

* Always create a new state object when adding an order (#1686)

Co-authored-by: Leandro Boscariol <leandro.boscariol@gnosis.io>

* Refactor: comparing account with order.owner lowercased for safety

Co-authored-by: Leandro Boscariol <leandro.boscariol@gnosis.io>

Commit:
Added new hook useTokenLazyNoMulticall to fetch token data one by one

Commit:
Using new hook to fetch tokens instead of multicall

Commit:
Merge remote-tracking branch 'origin/develop' into 1334/activity-orders-from-api-rebased

Commit:
Removed multicall/utils

Commit:
Removed useTokensLazy hook

Commit:
Renamed file useTokensLazy to useTokenLazy

Commit:
Renamed useTokenLazyNoMulticall to useTokenLazy

Commit:
Removed unused import

Commit:
Updated debug log messages

alfetopito commented:
Read for re-review @gnosis/gp-qa 

I've updated how the tokens are loaded, merged and resolved conflicts with latest `develop`.

To test, on mainnet place an order with this token `0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359` (SAI) or MKR.
You don't need to buy it, actually it's better if you cancel the trade because it's a "dead" token.
Example order with it https://protocol-explorer.dev.gnosisdev.com/orders/0x8524db28d6683e6c2996e158115a37e38e6d486a790f9ddebadbd8da114db9a7dd9eb88c5c6d2a85a08a96c7f0cccce27cb843cb6192e772

The special thing about SAI and MKR tokens is that their name/symbol are encoded as bytes32 which is different that 99.9% of the tokens.

Once you place the order, clear the token from the user added tokens and refresh the page and/or open in another browser.

Of course, testing with other regular tokens is important, not just the outliers.
The token should be loaded, added to user tokens and the order displayed on the activities list.

anxolin commented:
@alfetopito should we use a new branch "orders-api" as the base, so we can safely merge this even if its not complete? 

Maybe this would make easier to do small PR iterations after this one and we don't need to keep it open?

anxolin commented on src/custom/hooks/useTokenLazy.ts:
```diff
@@ -0,0 +1,115 @@
+      let name, symbol
+
+      // If no name or symbol, try the bytes32 contract
+      if (
```
The whole useEffect is extremely complex

I did #1857 #1858 to experiment a bit and try to refactor. Feel free to improve it here, or in other PR. In general my take aways are:
- Complex logic should be broken down into smaller pieces
- retry logic makes things a bit complex because of its signature viral nature (works very bad with chaining of retry logic)
- probably this libraries like SWR would help with these things
- ideally we would delay the creation of the contract for byte32 so we only create it if we need it

anxolin reviewed (commented):
Approve with comments. I think we can start consolidating, just to be able to reiterate on the small things in new PRs (this one is already biggg)

elena-zh commented:
Hey @alfetopito , SAI token was not added to the custom token list after its removal from there
https://watch.screencastify.com/v/efcFy6o6hcnlF5nWlkXo

However, orders history has been loaded

elena-zh commented:
During the day I was able to get 'Fee is insufficient' warning several times when I was testing #1861 PR. I thought that the issue is related to small trading amounts. 
But the same issue has appeared when I started testing the current PR, and trading amounts were high. I reported it here https://github.com/gnosis/cowswap/issues/1868#issuecomment-971757149

@alfetopito , could you please take a look at it?

elena-zh commented:
Also, I see 'token not found' message in the console when I place an order with a custom token (ZRX) in Rinkeby.
And it does not appear in the Custom token's list when I delete them, and then reload the page
![image](https://user-images.githubusercontent.com/70885163/142245388-1b5959c1-b9c1-4546-bb0c-126e67d9f266.png)
![image](https://user-images.githubusercontent.com/70885163/142245444-e254e6dc-55dc-4a45-ae36-24277031c4f2.png)

anxolin committed:
Small refactor in #1639 (#1857)

* Only init bytes32Contract if we really need to

* Fixing build error: only proceed if library is set

Co-authored-by: Leandro Boscariol <leandro.boscariol@gnosis.io>

Commit:
Merge remote-tracking branch 'origin/develop' into 1334/activity-orders-from-api-rebased

alfetopito commented:
@elena-zh could you try again again on #1906 ?

Just keep in mind it has the latest from `develop` but doesn't yet have all the fixes included on `release/1.5` branch, so some regressions are expected.

Also, I've tried on different browsers and several accounts/wallets and could not reproduce your issue.
Every time the tokens loaded fine.

:crossed_fingers: hopping this time it's fixed for you as well

elena-zh commented:
Hi @alfetopito , changes LGTM in #1906 
Besides, I reported a separate issue #1919  for displaying 'Filled on' on loaded activity list for 'Filled' orders, as API request might be changed and will have this info in the near future.

The most interesting thing (that is also a 'won't fix') is that activity in https://cowswap.dev.gnosisdev.com/#/swap shows a pending/cancelling/etc. order from https://barn.cowswap.exchange/#/swap as they both related to Dev server. 
The same will be for Prod and Stage environments.

alfetopito committed:
1334/with swr (#1906)

* Refactor: removed unnecessary variable

* Added SWR as dev dependency

* Added hook useApiOrders using useSWR hook

* Using new hook for fetching api orders

* Removed unused imports...

Co-authored-by: Leandro Boscariol <leandro.boscariol@gnosis.io>

Commit:
Merge remote-tracking branch 'origin/develop' into 1334/activity-orders-from-api-rebased

Commit:
Merge remote-tracking branch 'origin/develop' into 1334/activity-orders-from-api-rebased

anxolin commented on src/custom/api/gnosisProtocol/hooks.ts:
```diff
@@ -0,0 +1,15 @@
+export function useApiOrders(account?: string | null): OrderMetaData[] | undefined {
+  const { chainId } = useActiveWeb3React()
+
+  const { data } = useSWR<OrderMetaData[]>(['orders', account, chainId], () =>
```
How much would this be caching? 

Should this have something like `refreshInterval`?

alfetopito replied:
We definitely can https://swr.vercel.app/docs/revalidation

I started without it to make it simpler and avoid potential race conditions.

How often do you think we should refresh it?

The main reason why we'd want to refresh is in case there is more than one instance of the page open in different browsers/devices.

In my view this is not a common use case to to need this.
Anyway, I'll set it to `30s` for testing and we can re-evaluate.

alfetopito replied:
https://github.com/gnosis/cowswap/pull/1951

anxolin commented on src/custom/hooks/useTokenLazy.ts:
```diff
@@ -0,0 +1,148 @@
+  return bytes32ContractsCache[address] || getBytes32TokenContract(address, undefined, library, account, chainId)
+}
+
+async function _getTokenInfo(params: GetTokenInfoParams): Promise<TokenInfo | null> {
```
this one i feel is too big, but i think we can leave it like this for now...
I also want this PR to be merged :)

anxolin commented on src/custom/state/orders/updaters/ApiOrdersUpdater.ts:
```diff
@@ -0,0 +1,188 @@
+  const tokensAreLoaded = useMemo(() => Object.keys(allTokens).length > 0, [allTokens])
+  const addOrUpdateOrders = useAddOrUpdateOrders()
+  const getToken = useTokenLazy()
+  const apiOrders = useApiOrders(account)
```
nit: the name `apiOrders` would make me think this is an API for getting orders, but is just orders from an API
Also, calling our endpoints just API can be confusing. Maybe we give another name?

anxolin reviewed (approved):
Looks great!

nit: maybe we want to consider to change the message to "No activity yet". Now we are sure there's no activity for that account

![image](https://user-images.githubusercontent.com/2352112/143934030-5e6d2c7b-c4b7-4747-bbcb-3b703fcf740d.png)


Why is this PR adding `presignaturePending` all over? isn't this from presign PRs already?

alfetopito commented:
@anxolin I'll merge as is and address your comments in follow up PRs

> nit: maybe we want to consider to change the message to "No activity yet". Now we are sure there's no activity for that account

Well there might be approvals/wraps. We are not syncing them yet

> Why is this PR adding presignaturePending all over? isn't this from presign PRs already?

This type was not exposed where I needed it
